File: merge-tonic-lib/tonic/hydration/persistence.py
"""Persistence safety helpers for vendored Chroma stores."""

from __future__ import annotations

from dataclasses import asdict, dataclass
from datetime import UTC, datetime
import hashlib
import json
from pathlib import Path
from uuid import uuid4

from .errors import HydrationPersistError
from .paths import DEFAULT_INDEX_STATE_FILE_NAME
from .types import HYDRATION_PIPELINE_VERSION

HYDRATION_PERSIST_MANIFEST_SCHEMA = "tonic-hydration-persist-manifest"
HYDRATION_PERSIST_MANIFEST_FILE_NAME = "persist-manifest.json"
HYDRATION_PERSIST_LOCK_FILE_NAME = ".writer.lock"
HYDRATION_CACHE_KEY_SCHEMA_VERSION = "1"


@dataclass(frozen=True)
class HydrationPersistManifest:
    persist_root: str
    index_state_file: str
    writer_kind: str
    entries: list[str]
    schema: str = HYDRATION_PERSIST_MANIFEST_SCHEMA
    pipeline_version: str = HYDRATION_PIPELINE_VERSION
    cache_key: str = ""
    updated_at: str = ""


@dataclass(frozen=True)
class HydrationPersistHealth:
    ok: bool
    cold_start: bool
    has_index_state: bool
    has_store_files: bool
    entries: list[str]
    manifest_path: str
    reason: str = ""
    missing_entries: list[str] | None = None


@dataclass(frozen=True)
class HydrationCacheKey:
    cache_key: str
    pr_list_hash: str
    payload: dict[str, str]


class HydrationPersistLock:
    def __init__(self, lock_path: Path, token: str) -> None:
        self.lock_path = lock_path
        self.token = token

    def release(self) -> None:
        if not self.lock_path.is_file():
            return
        try:
            current = json.loads(self.lock_path.read_text(encoding="utf-8"))
        except Exception:
            return
        if current.get("token") != self.token:
            return
        self.lock_path.unlink(missing_ok=True)


def build_hydration_cache_key(
    *,
    strategy_id: str,
    pr_identifiers: list[str] | None,
    embedder_model: str,
    chunker_version: str,
    scope: str = "",
    prompt_profile: str = "",
    dry_run: bool = False,
    historical_since: str = "",
    historical_base_ref: str = "",
    historical_state: str = "",
) -> HydrationCacheKey:
    normalized_prs = sorted([value.strip() for value in (pr_identifiers or []) if value.strip()])
    pr_list_hash = hashlib.sha256("\n".join(normalized_prs).encode("utf-8")).hexdigest()
    payload = {
        "schema_version": HYDRATION_CACHE_KEY_SCHEMA_VERSION,
        "strategy_id": strategy_id,
        "pr_list_hash": pr_list_hash,
        "embedder_model": embedder_model,
        "chunker_version": chunker_version,
        "scope": scope.strip(),
        "prompt_profile": prompt_profile.strip(),
        "dry_run": "true" if dry_run else "false",
        "historical_since": historical_since.strip(),
        "historical_base_ref": historical_base_ref.strip(),
        "historical_state": historical_state.strip(),
    }
    return HydrationCacheKey(
        cache_key=hashlib.sha256(json.dumps(payload, sort_keys=True).encode("utf-8")).hexdigest(),
        pr_list_hash=pr_list_hash,
        payload=payload,
    )


def resolve_hydration_persist_manifest_path(persist_root: str | Path) -> Path:
    return Path(persist_root) / HYDRATION_PERSIST_MANIFEST_FILE_NAME


def load_hydration_persist_manifest(persist_root: str | Path) -> HydrationPersistManifest | None:
    manifest_path = resolve_hydration_persist_manifest_path(persist_root)
    if not manifest_path.is_file():
        return None
    data = json.loads(manifest_path.read_text(encoding="utf-8"))
    return HydrationPersistManifest(**data)


def write_hydration_persist_manifest(
    *,
    persist_root: str | Path,
    writer_kind: str,
    cache_key: str = "",
    index_state_file: str = DEFAULT_INDEX_STATE_FILE_NAME,
) -> HydrationPersistManifest:
    root = Path(persist_root)
    root.mkdir(parents=True, exist_ok=True)
    entries = sorted([item.name for item in root.iterdir() if item.name != HYDRATION_PERSIST_LOCK_FILE_NAME])
    manifest = HydrationPersistManifest(
        persist_root=str(root),
        index_state_file=index_state_file,
        writer_kind=writer_kind,
        cache_key=cache_key,
        entries=entries,
        updated_at=datetime.now(UTC).isoformat(),
    )
    resolve_hydration_persist_manifest_path(root).write_text(
        json.dumps(asdict(manifest), indent=2) + "\n",
        encoding="utf-8",
    )
    return manifest


def check_hydration_persist_health(persist_root: str | Path) -> HydrationPersistHealth:
    root = Path(persist_root)
    manifest_path = resolve_hydration_persist_manifest_path(root)
    if not root.exists():
        return HydrationPersistHealth(
            ok=True,
            cold_start=True,
            has_index_state=False,
            has_store_files=False,
            entries=[],
            manifest_path=str(manifest_path),
        )
    entries = sorted([item.name for item in root.iterdir()])
    if not entries:
        return HydrationPersistHealth(
            ok=True,
            cold_start=True,
            has_index_state=False,
            has_store_files=False,
            entries=entries,
            manifest_path=str(manifest_path),
        )
    has_index_state = DEFAULT_INDEX_STATE_FILE_NAME in entries
    has_store_files = any(
        name
        not in {
            DEFAULT_INDEX_STATE_FILE_NAME,
            HYDRATION_PERSIST_MANIFEST_FILE_NAME,
            HYDRATION_PERSIST_LOCK_FILE_NAME,
        }
        for name in entries
    )
    if has_store_files and not has_index_state:
        return HydrationPersistHealth(
            ok=False,
            cold_start=False,
            has_index_state=has_index_state,
            has_store_files=has_store_files,
            entries=entries,
            manifest_path=str(manifest_path),
            reason="Persist root has Chroma files but no index-state.json. Restore the full tree or rebuild the cache.",
        )
    manifest = load_hydration_persist_manifest(root)
    if manifest is None:
        ok = not has_store_files and not has_index_state
        return HydrationPersistHealth(
            ok=ok,
            cold_start=ok,
            has_index_state=has_index_state,
            has_store_files=has_store_files,
            entries=entries,
            manifest_path=str(manifest_path),
            reason=""
            if ok
            else "Persist root is missing persist-manifest.json. Rebuild the persisted Chroma tree instead of restoring partial files.",
        )
    missing_entries = [entry for entry in manifest.entries if not (root / entry).exists()]
    if missing_entries:
        return HydrationPersistHealth(
            ok=False,
            cold_start=False,
            has_index_state=has_index_state,
            has_store_files=has_store_files,
            entries=entries,
            manifest_path=str(manifest_path),
            reason=f"Persist root is missing required entries from {HYDRATION_PERSIST_MANIFEST_FILE_NAME}. Restore the full tree or rebuild the cache.",
            missing_entries=missing_entries,
        )
    return HydrationPersistHealth(
        ok=True,
        cold_start=False,
        has_index_state=has_index_state,
        has_store_files=has_store_files,
        entries=entries,
        manifest_path=str(manifest_path),
    )


def assert_hydration_persist_health(persist_root: str | Path) -> HydrationPersistHealth:
    health = check_hydration_persist_health(persist_root)
    if not health.ok:
        raise HydrationPersistError(health.reason or "Hydration persist root is not safe to reuse.")
    return health


def acquire_hydration_persist_lock(persist_root: str | Path, owner: str) -> HydrationPersistLock:
    root = Path(persist_root)
    root.mkdir(parents=True, exist_ok=True)
    lock_path = root / HYDRATION_PERSIST_LOCK_FILE_NAME
    token = str(uuid4())
    payload = {
        "owner": owner,
        "token": token,
        "pid": str(__import__("os").getpid()),
        "acquired_at": datetime.now(UTC).isoformat(),
    }
    try:
        with lock_path.open("x", encoding="utf-8") as handle:
            handle.write(json.dumps(payload, indent=2) + "\n")
    except FileExistsError as exc:
        detail = lock_path.read_text(encoding="utf-8").strip() if lock_path.is_file() else ""
        raise HydrationPersistError(
            f"Hydration persist root is already claimed by another writer at {lock_path}.{f' Existing lock: {detail}' if detail else ''}"
        ) from exc
    return HydrationPersistLock(lock_path, token)

File: packages/tonic-core/src/hydration/persistence.ts
import { createHash, randomUUID } from "node:crypto";
import * as fs from "node:fs";
import * as path from "node:path";

import { HYDRATION_PIPELINE_VERSION } from "./types";
import { DEFAULT_INDEX_STATE_FILE_NAME } from "./paths";
import { HydrationPersistError } from "./errors";

export const HYDRATION_PERSIST_MANIFEST_SCHEMA = "tonic-hydration-persist-manifest";
export const HYDRATION_PERSIST_MANIFEST_FILE_NAME = "persist-manifest.json";
export const HYDRATION_PERSIST_LOCK_FILE_NAME = ".writer.lock";
export const HYDRATION_CACHE_KEY_SCHEMA_VERSION = "1";

export type HydrationPersistManifest = {
  schema: typeof HYDRATION_PERSIST_MANIFEST_SCHEMA;
  pipeline_version: string;
  persist_root: string;
  index_state_file: string;
  writer_kind: string;
  cache_key?: string;
  entries: string[];
  updated_at: string;
};

export type HydrationPersistHealth = {
  ok: boolean;
  coldStart: boolean;
  hasIndexState: boolean;
  hasStoreFiles: boolean;
  entries: string[];
  manifestPath: string;
  reason?: string;
  missingEntries?: string[];
};

export type HydrationCacheKeyParts = {
  strategyId: string;
  prIdentifiers?: string[];
  embedderModel: string;
  chunkerVersion: string;
  scope?: string;
  promptProfile?: string;
  dryRun?: boolean;
  historicalSince?: string;
  historicalBaseRef?: string;
  historicalState?: "merged" | "open" | "all";
};

export type HydrationCacheKey = {
  cacheKey: string;
  prListHash: string;
  payload: {
    schema_version: string;
    strategy_id: string;
    pr_list_hash: string;
    embedder_model: string;
    chunker_version: string;
    scope: string;
    prompt_profile: string;
    dry_run: "true" | "false";
    historical_since: string;
    historical_base_ref: string;
    historical_state: string;
  };
};

export type HydrationPersistLock = {
  lockPath: string;
  token: string;
  release(): void;
};

function normalizePrIdentifiers(values: string[] | undefined): string[] {
  return [...(values ?? [])].map((value) => value.trim()).filter(Boolean).sort();
}

function listPersistEntries(persistRoot: string): string[] {
  if (!fs.existsSync(persistRoot)) {
    return [];
  }
  return fs.readdirSync(persistRoot).sort();
}

function hasStoreFiles(entries: string[]): boolean {
  return entries.some((entry) =>
    ![
      DEFAULT_INDEX_STATE_FILE_NAME,
      HYDRATION_PERSIST_MANIFEST_FILE_NAME,
      HYDRATION_PERSIST_LOCK_FILE_NAME,
    ].includes(entry),
  );
}

export function buildHydrationCacheKey(parts: HydrationCacheKeyParts): HydrationCacheKey {
  const prIdentifiers = normalizePrIdentifiers(parts.prIdentifiers);
  const prListHash = createHash("sha256").update(prIdentifiers.join("\n"), "utf8").digest("hex");
  const payload = {
    schema_version: HYDRATION_CACHE_KEY_SCHEMA_VERSION,
    strategy_id: parts.strategyId,
    pr_list_hash: prListHash,
    embedder_model: parts.embedderModel,
    chunker_version: parts.chunkerVersion,
    scope: (parts.scope ?? "").trim(),
    prompt_profile: (parts.promptProfile ?? "").trim(),
    dry_run: parts.dryRun ? "true" as const : "false" as const,
    historical_since: (parts.historicalSince ?? "").trim(),
    historical_base_ref: (parts.historicalBaseRef ?? "").trim(),
    historical_state: (parts.historicalState ?? "").trim(),
  };
  return {
    cacheKey: createHash("sha256").update(JSON.stringify(payload), "utf8").digest("hex"),
    prListHash,
    payload,
  };
}

export function resolveHydrationPersistManifestPath(persistRoot: string): string {
  return path.join(persistRoot, HYDRATION_PERSIST_MANIFEST_FILE_NAME);
}

export function writeHydrationPersistManifest(params: {
  persistRoot: string;
  writerKind: string;
  cacheKey?: string;
  indexStateFile?: string;
}): HydrationPersistManifest {
  fs.mkdirSync(params.persistRoot, { recursive: true });
  const manifestPath = resolveHydrationPersistManifestPath(params.persistRoot);
  const entries = listPersistEntries(params.persistRoot).filter((entry) => entry !== HYDRATION_PERSIST_LOCK_FILE_NAME);
  const manifest: HydrationPersistManifest = {
    schema: HYDRATION_PERSIST_MANIFEST_SCHEMA,
    pipeline_version: HYDRATION_PIPELINE_VERSION,
    persist_root: params.persistRoot,
    index_state_file: params.indexStateFile ?? DEFAULT_INDEX_STATE_FILE_NAME,
    writer_kind: params.writerKind,
    cache_key: params.cacheKey,
    entries,
    updated_at: new Date().toISOString(),
  };
  fs.writeFileSync(manifestPath, JSON.stringify(manifest, null, 2) + "\n", "utf8");
  return manifest;
}

export function loadHydrationPersistManifest(persistRoot: string): HydrationPersistManifest | null {
  const manifestPath = resolveHydrationPersistManifestPath(persistRoot);
  if (!fs.existsSync(manifestPath)) {
    return null;
  }
  return JSON.parse(fs.readFileSync(manifestPath, "utf8")) as HydrationPersistManifest;
}

export function checkHydrationPersistHealth(persistRoot: string): HydrationPersistHealth {
  const manifestPath = resolveHydrationPersistManifestPath(persistRoot);
  const entries = listPersistEntries(persistRoot);
  const indexStateFile = DEFAULT_INDEX_STATE_FILE_NAME;
  const hasIndexState = entries.includes(indexStateFile);
  const storeFilesPresent = hasStoreFiles(entries);
  if (entries.length === 0) {
    return {
      ok: true,
      coldStart: true,
      hasIndexState: false,
      hasStoreFiles: false,
      entries,
      manifestPath,
    };
  }
  if (storeFilesPresent && !hasIndexState) {
    return {
      ok: false,
      coldStart: false,
      hasIndexState,
      hasStoreFiles: storeFilesPresent,
      entries,
      manifestPath,
      reason: "Persist root has Chroma files but no index-state.json. Restore the full tree or rebuild the cache.",
    };
  }
  const manifest = loadHydrationPersistManifest(persistRoot);
  if (!manifest) {
    return {
      ok: !storeFilesPresent && !hasIndexState,
      coldStart: !storeFilesPresent && !hasIndexState,
      hasIndexState,
      hasStoreFiles: storeFilesPresent,
      entries,
      manifestPath,
      reason:
        storeFilesPresent || hasIndexState ?
          "Persist root is missing persist-manifest.json. Rebuild the persisted Chroma tree instead of restoring partial files."
        : undefined,
    };
  }
  const missingEntries = manifest.entries.filter((entry) => !fs.existsSync(path.join(persistRoot, entry)));
  if (missingEntries.length > 0) {
    return {
      ok: false,
      coldStart: false,
      hasIndexState,
      hasStoreFiles: storeFilesPresent,
      entries,
      manifestPath,
      missingEntries,
      reason: `Persist root is missing required entries from ${HYDRATION_PERSIST_MANIFEST_FILE_NAME}. Restore the full tree or rebuild the cache.`,
    };
  }
  return {
    ok: true,
    coldStart: false,
    hasIndexState,
    hasStoreFiles: storeFilesPresent,
    entries,
    manifestPath,
  };
}

export function assertHydrationPersistHealth(persistRoot: string): HydrationPersistHealth {
  const health = checkHydrationPersistHealth(persistRoot);
  if (!health.ok) {
    throw new HydrationPersistError(health.reason ?? "Hydration persist root is not safe to reuse.");
  }
  return health;
}

export function acquireHydrationPersistLock(persistRoot: string, owner: string): HydrationPersistLock {
  fs.mkdirSync(persistRoot, { recursive: true });
  const lockPath = path.join(persistRoot, HYDRATION_PERSIST_LOCK_FILE_NAME);
  const token = randomUUID();
  const payload = {
    owner,
    token,
    pid: process.pid,
    acquired_at: new Date().toISOString(),
  };
  try {
    const fd = fs.openSync(lockPath, "wx");
    try {
      fs.writeFileSync(fd, JSON.stringify(payload, null, 2) + "\n", "utf8");
    } finally {
      fs.closeSync(fd);
    }
  } catch {
    let detail = "";
    if (fs.existsSync(lockPath)) {
      detail = fs.readFileSync(lockPath, "utf8").trim();
    }
    throw new HydrationPersistError(
      `Hydration persist root is already claimed by another writer at ${lockPath}.${detail ? ` Existing lock: ${detail}` : ""}`,
    );
  }
  return {
    lockPath,
    token,
    release(): void {
      if (!fs.existsSync(lockPath)) {
        return;
      }
      try {
        const current = JSON.parse(fs.readFileSync(lockPath, "utf8")) as { token?: string };
        if (current.token !== token) {
          return;
        }
      } catch {
        return;
      }
      fs.unlinkSync(lockPath);
    },
  };
}
